test_conformance/extensions/cl_khr_command_buffer/negative_command_buffer_copy.cpp

```diff
@@ -0,0 +1,506 @@
+                               "CL_INVALID_COMMAND_BUFFER_KHR",
+                               TEST_FAIL);
+
+        const cl_image_format formats = { CL_RGBA, CL_UNSIGNED_INT8 };
+        clMemWrapper image = create_image_2d(
+            context, CL_MEM_READ_WRITE, &formats, 512, 512, 0, NULL, &error);
+        test_error(error, "create_image_2d failed");
+        clMemWrapper buffer = clCreateBuffer(context, CL_MEM_READ_WRITE,
+                                             data_size(), nullptr, &error);
+        test_error(error, "Unable to create buffer");
+
+        error = clCommandCopyImageToBufferKHR(nullptr, nullptr, image, buffer,
+                                              origin, region, 0, 0, nullptr,
+                                              nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_COMMAND_BUFFER_KHR,
+                               "clCommandCopyImageToBufferKHR should return "
+                               "CL_INVALID_COMMAND_BUFFER_KHR",
+                               TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+};
+
+// CL_INVALID_OPERATION if command_buffer has been finalized.
+struct CommandBufferCopyFinalizedCommandBuffer : public BasicCommandBufferTest
+{
+    using BasicCommandBufferTest::BasicCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_int error = clFinalizeCommandBufferKHR(command_buffer);
+        test_error(error, "clFinalizeCommandBufferKHR failed");
+
+        error = clCommandCopyBufferKHR(command_buffer, nullptr, in_mem, out_mem,
+                                       0, 0, data_size(), 0, nullptr, nullptr,
+                                       nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_OPERATION,
+                               "clCommandCopyBufferKHR should return "
+                               "CL_INVALID_OPERATION",
+                               TEST_FAIL);
+
+        const size_t origin[3] = { 0, 0, 0 }, region[3] = { 512, 512, 1 };
+
+        error = clCommandCopyBufferRectKHR(
+            command_buffer, nullptr, in_mem, out_mem, origin, origin, region, 0,
+            0, 0, 0, 0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(error, CL_INVALID_OPERATION,
+                               "clCommandCopyBufferRectKHR should return "
+                               "CL_INVALID_OPERATION",
+                               TEST_FAIL);
+
+        const cl_image_format formats = { CL_RGBA, CL_UNSIGNED_INT8 };
+        clMemWrapper image = create_image_2d(
+            context, CL_MEM_READ_WRITE, &formats, 512, 512, 0, NULL, &error);
+        test_error(error, "create_image_2d failed");
+        clMemWrapper buffer = clCreateBuffer(context, CL_MEM_READ_WRITE,
+                                             data_size(), nullptr, &error);
+        test_error(error, "Unable to create buffer");
+
+        error = clCommandCopyImageToBufferKHR(nullptr, nullptr, image, buffer,
```

Comment:
I think this needs to pass the command buffer here instead of `nullptr`, otherwise the return value will be `CL_INVALID_COMMAND_BUFFER_KHR`, not `CL_INVALID_OPERATION`.

```suggestion
        error = clCommandCopyImageToBufferKHR(command_buffer, nullptr, image, buffer,
```

Reply:
@bashbaug corrected